StreamingOutput: Cannot reserve 65536 bytes of direct buffer memory (allocated: 760195093, limit: 760217600)

### Describe the bug

hello,

in quarkus 3.15.6.1, I got the following controller:

```java    
    @Blocking
    @GET
    @Path("download")
    @Produces(MediaType.APPLICATION_OCTET_STREAM)
    public Response download(@QueryParam("directory") String directory, @QueryParam("file") String filePath) throws IOException {
        java.nio.file.Path root = getDirectory(directory).normalize().toRealPath();
        java.nio.file.Path finalPath = root.resolve(filePath).normalize().toRealPath();

        // security check: avoid attackers to access arbitrary files in the filesystem
        if (!finalPath.startsWith(root)) {
            return Response.status(Response.Status.NOT_FOUND).build();
        }

        String fileName = String.valueOf(finalPath.getFileName());
    
        StreamingOutput responseBody = outputStream -> streamFile(finalPath, outputStream);
        return Response.ok(responseBody)
                .header("Content-Disposition", "attachment; filename=\"" + fileName + "\"")
                .build();
    }

    private static void streamFile(java.nio.file.Path file, OutputStream outputStream) throws IOException {
        try (InputStream inputStream = new BufferedInputStream(Files.newInputStream(file))) {
            inputStream.transferTo(outputStream);
        }
    }
```

It works great for small files but it fails when I'm trying to download a 6GB file:


IIRC, in previous version of quarkus, this method was able to download big files but I cannot remember exactly which version.  This happens on Linux... now I'm going to test on Windows 11.



### Expected behavior

A StreamingOutput with transferTo downloads the file successfully.

### Actual behavior


```
2025-09-29 05:48:03,677 ERROR [io.qua.ver.htt.run.QuarkusErrorHandler] (executor-thread-1) HTTP Request to /utilities/filesystem/download?directory=yourkit&file=dump-cdm-api-blue-5f7d5f5544-4hlc5-2025-09-16_15-04-58%27 failed, error id: 65ab25d9-f8fc-46d1-a205-53b7398ed874-3 MDC={} : java.lang.RuntimeException: java.lang.OutOfMemoryError: Cannot reserve 65536 bytes of direct buffer memory (allocated: 760188389, limit: 760217600)
at org.jboss.resteasy.reactive.server.core.ServerSerialisers.invokeWriter(ServerSerialisers.java:256)
at org.jboss.resteasy.reactive.server.core.serialization.DynamicEntityWriter.write(DynamicEntityWriter.java:119)
at org.jboss.resteasy.reactive.server.handlers.ResponseWriterHandler.handle(ResponseWriterHandler.java:32)
at io.quarkus.resteasy.reactive.server.runtime.QuarkusResteasyReactiveRequestContext.invokeHandler(QuarkusResteasyReactiveRequestContext.java:147)
at org.jboss.resteasy.reactive.common.core.AbstractResteasyReactiveContext.run(AbstractResteasyReactiveContext.java:147)
at io.quarkus.vertx.core.runtime.VertxCoreRecorder$14.runWith(VertxCoreRecorder.java:635)
at org.jboss.threads.EnhancedQueueExecutor$Task.doRunWith(EnhancedQueueExecutor.java:2516)
at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2495)
at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1521)
at org.jboss.threads.DelegatingRunnable.run(DelegatingRunnable.java:11)
at org.jboss.threads.ThreadLocalResettingRunnable.run(ThreadLocalResettingRunnable.java:11)
at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
at java.base/java.lang.Thread.run(Thread.java:1583)
Caused by: java.lang.OutOfMemoryError: Cannot reserve 65536 bytes of direct buffer memory (allocated: 760188389, limit: 760217600)
at java.base/java.nio.Bits.reserveMemory(Bits.java:178)
at java.base/java.nio.DirectByteBuffer.<init>(DirectByteBuffer.java:111)
at java.base/java.nio.ByteBuffer.allocateDirect(ByteBuffer.java:360)
at io.netty.buffer.PoolArena$DirectArena.allocateDirect(PoolArena.java:718)
at io.netty.buffer.PoolArena$DirectArena.newChunk(PoolArena.java:693)
at io.netty.buffer.PoolArena.allocateNormal(PoolArena.java:213)
at io.netty.buffer.PoolArena.tcacheAllocateSmall(PoolArena.java:178)
at io.netty.buffer.PoolArena.allocate(PoolArena.java:135)
at io.netty.buffer.PoolArena.allocate(PoolArena.java:127)
at io.netty.buffer.PooledByteBufAllocator.newDirectBuffer(PooledByteBufAllocator.java:403)
at io.netty.buffer.AbstractByteBufAllocator.directBuffer(AbstractByteBufAllocator.java:188)
at io.netty.buffer.AbstractByteBufAllocator.directBuffer(AbstractByteBufAllocator.java:179)
at io.quarkus.vertx.utils.AppendBuffer.append(AppendBuffer.java:136)
at io.quarkus.vertx.utils.VertxOutputStream.write(VertxOutputStream.java:178)
at java.base/java.io.InputStream.transferTo(InputStream.java:797)
at java.base/sun.nio.ch.ChannelInputStream.transferTo(ChannelInputStream.java:268)
at java.base/java.io.BufferedInputStream.implTransferTo(BufferedInputStream.java:652)
at java.base/java.io.BufferedInputStream.transferTo(BufferedInputStream.java:631)
at com.sixgroup.flex.cdmstore.management.utilities.filesystem.FilesystemController.streamFile(FilesystemController.java:116)
at com.sixgroup.flex.cdmstore.management.utilities.filesystem.FilesystemController.lambda$download$0(FilesystemController.java:108)
at org.jboss.resteasy.reactive.server.providers.serialisers.StreamingOutputMessageBodyWriter.writeResponse(StreamingOutputMessageBodyWriter.java:49)
at org.jboss.resteasy.reactive.server.providers.serialisers.StreamingOutputMessageBodyWriter.writeResponse(StreamingOutputMessageBodyWriter.java:17)
at org.jboss.resteasy.reactive.server.core.ServerSerialisers.invokeWriter(ServerSerialisers.java:218)
... 12 more
```

The side effect is that the file is truncated at different positions (sometimes after 1.7GB, sometimes after 1GB) and it is not useable anymore.

### How to Reproduce?

_No response_

### Output of `uname -a` or `ver`

_No response_

### Output of `java -version`

openjdk version "21.0.7" 2025-04-15 LTS OpenJDK Runtime Environment Zulu21.42+20-SA (build 21.0.7+6-LTS) OpenJDK 64-Bit Server VM Zulu21.42+20-SA (build 21.0.7+6-LTS, mixed mode, sharing)

### Quarkus version or git rev

3.15.6.1

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

_No response_

### Additional information

_No response_

since it is possible to reproduce it on my local machine, I tried to find a temp workaround:
1) reimplemented the loop of `transferTo`, something like:
```java
    private static void streamFile(java.nio.file.Path file, OutputStream outputStream) throws IOException {
        final int bufferSize = 16384;
        try (InputStream inputStream = Files.newInputStream(file)) {
            long transferred = 0;
            byte[] buffer = new byte[bufferSize];
            int read;
            while ((read = inputStream.read(buffer)) >= 0) {
                outputStream.write(buffer, 0, read);
                outputStream.flush();
                transferred += read;
            }
            LOGGER.info("successfully downloaded {} (size {})", file, transferred);
        }
    }
```

2) let quarkus to the job completely:
```java
        return Response.ok(Files.newInputStream(finalPath))
                .header("Content-Disposition", "attachment; filename=\"" + fileName + "\"")
                .build();
 ```
3) same as 2) but without @Blocking annotation (not a good idea because there are some blocking calls but I wanted to try anyways)

Interestingly in all cases I got same exception, like this:
```
2025-09-29 09:02:00,217 ERROR [io.qua.ver.htt.run.QuarkusErrorHandler] (executor-thread-1) HTTP Request to /utilities/filesystem/download?directory=yourkit&file=test_6000MB.bin failed, error id: 9657146c-0b69-44b7-a4b9-88b16227581d-1 MDC={} : java.lang.RuntimeException: java.lang.OutOfMemoryError: Cannot reserve 65536 bytes of direct buffer memory (allocated: 786380409, limit: 786432000)
	at org.jboss.resteasy.reactive.server.core.ServerSerialisers.invokeWriter(ServerSerialisers.java:256)
	at org.jboss.resteasy.reactive.server.core.serialization.DynamicEntityWriter.write(DynamicEntityWriter.java:119)
	at org.jboss.resteasy.reactive.server.handlers.ResponseWriterHandler.handle(ResponseWriterHandler.java:32)
	at io.quarkus.resteasy.reactive.server.runtime.QuarkusResteasyReactiveRequestContext.invokeHandler(QuarkusResteasyReactiveRequestContext.java:147)
	at org.jboss.resteasy.reactive.common.core.AbstractResteasyReactiveContext.run(AbstractResteasyReactiveContext.java:147)
	at io.quarkus.vertx.core.runtime.VertxCoreRecorder$14.runWith(VertxCoreRecorder.java:635)
	at org.jboss.threads.EnhancedQueueExecutor$Task.doRunWith(EnhancedQueueExecutor.java:2516)
	at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2495)
	at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1521)
	at org.jboss.threads.DelegatingRunnable.run(DelegatingRunnable.java:11)
	at org.jboss.threads.ThreadLocalResettingRunnable.run(ThreadLocalResettingRunnable.java:11)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.base/java.lang.Thread.run(Thread.java:1583)
Caused by: java.lang.OutOfMemoryError: Cannot reserve 65536 bytes of direct buffer memory (allocated: 786380409, limit: 786432000)
	at java.base/java.nio.Bits.reserveMemory(Bits.java:178)
	at java.base/java.nio.DirectByteBuffer.<init>(DirectByteBuffer.java:111)
	at java.base/java.nio.ByteBuffer.allocateDirect(ByteBuffer.java:360)
	at io.netty.buffer.PoolArena$DirectArena.allocateDirect(PoolArena.java:718)
	at io.netty.buffer.PoolArena$DirectArena.newChunk(PoolArena.java:693)
	at io.netty.buffer.PoolArena.allocateNormal(PoolArena.java:213)
	at io.netty.buffer.PoolArena.tcacheAllocateSmall(PoolArena.java:178)
	at io.netty.buffer.PoolArena.allocate(PoolArena.java:135)
	at io.netty.buffer.PoolArena.allocate(PoolArena.java:127)
	at io.netty.buffer.PooledByteBufAllocator.newDirectBuffer(PooledByteBufAllocator.java:403)
	at io.netty.buffer.AbstractByteBufAllocator.directBuffer(AbstractByteBufAllocator.java:188)
	at io.netty.buffer.AbstractByteBufAllocator.directBuffer(AbstractByteBufAllocator.java:179)
	at io.quarkus.vertx.utils.AppendBuffer.append(AppendBuffer.java:136)
	at io.quarkus.vertx.utils.VertxOutputStream.write(VertxOutputStream.java:178)
	at java.base/java.io.InputStream.transferTo(InputStream.java:797)
	at java.base/sun.nio.ch.ChannelInputStream.transferTo(ChannelInputStream.java:268)
	at org.jboss.resteasy.reactive.common.providers.serialisers.InputStreamMessageBodyHandler.writeTo(InputStreamMessageBodyHandler.java:38)
	at org.jboss.resteasy.reactive.server.providers.serialisers.ServerInputStreamMessageBodyHandler.writeResponse(ServerInputStreamMessageBodyHandler.java:46)
	at org.jboss.resteasy.reactive.server.providers.serialisers.ServerInputStreamMessageBodyHandler.writeResponse(ServerInputStreamMessageBodyHandler.java:17)
	at org.jboss.resteasy.reactive.server.core.ServerSerialisers.invokeWriter(ServerSerialisers.java:218)
	... 12 more
```

> I don't see `StreamingOutput` being used in the reproducer.
ops, you're right: 
[quarkus-issue-50317-StreamingOutput.zip](https://github.com/user-attachments/files/22612618/quarkus-issue-50317-StreamingOutput.zip)

Here is the stracktrace for the `StreamingOutput` version
```
Caused by: java.lang.OutOfMemoryError: Cannot reserve 65536 bytes of direct buffer memory (allocated: 268428896, limit: 268435456)
	at java.base/java.nio.Bits.reserveMemory(Bits.java:178)
	at java.base/java.nio.DirectByteBuffer.<init>(DirectByteBuffer.java:111)
	at java.base/java.nio.ByteBuffer.allocateDirect(ByteBuffer.java:360)
	at io.netty.buffer.PoolArena$DirectArena.allocateDirect(PoolArena.java:718)
	at io.netty.buffer.PoolArena$DirectArena.newChunk(PoolArena.java:693)
	at io.netty.buffer.PoolArena.allocateNormal(PoolArena.java:213)
	at io.netty.buffer.PoolArena.tcacheAllocateSmall(PoolArena.java:178)
	at io.netty.buffer.PoolArena.allocate(PoolArena.java:135)
	at io.netty.buffer.PoolArena.allocate(PoolArena.java:127)
	at io.netty.buffer.PooledByteBufAllocator.newDirectBuffer(PooledByteBufAllocator.java:403)
	at io.netty.buffer.AbstractByteBufAllocator.directBuffer(AbstractByteBufAllocator.java:188)
	at io.netty.buffer.AbstractByteBufAllocator.directBuffer(AbstractByteBufAllocator.java:179)
	at io.quarkus.vertx.utils.AppendBuffer.append(AppendBuffer.java:136)
	at io.quarkus.vertx.utils.VertxOutputStream.write(VertxOutputStream.java:178)
	at java.base/java.io.InputStream.transferTo(InputStream.java:797)
	at java.base/sun.nio.ch.ChannelInputStream.transferTo(ChannelInputStream.java:268)
	at java.base/java.io.BufferedInputStream.implTransferTo(BufferedInputStream.java:652)
	at java.base/java.io.BufferedInputStream.transferTo(BufferedInputStream.java:631)
	at org.example.DownloadController.streamFile(DownloadController.java:44)
	at org.example.DownloadController.lambda$download$0(DownloadController.java:36)
	at org.jboss.resteasy.reactive.server.providers.serialisers.StreamingOutputMessageBodyWriter.writeResponse(StreamingOutputMessageBodyWriter.java:49)
	at org.jboss.resteasy.reactive.server.providers.serialisers.StreamingOutputMessageBodyWriter.writeResponse(StreamingOutputMessageBodyWriter.java:17)
	at org.jboss.resteasy.reactive.server.core.ServerSerialisers.invokeWriter(ServerSerialisers.java:218)
	... 12 more
```

And here is the stacktrace for the `InputStream` version:
```
Caused by: java.lang.OutOfMemoryError: Cannot reserve 65536 bytes of direct buffer memory (allocated: 268433876, limit: 268435456)
	at java.base/java.nio.Bits.reserveMemory(Bits.java:178)
	at java.base/java.nio.DirectByteBuffer.<init>(DirectByteBuffer.java:111)
	at java.base/java.nio.ByteBuffer.allocateDirect(ByteBuffer.java:360)
	at io.netty.buffer.PoolArena$DirectArena.allocateDirect(PoolArena.java:718)
	at io.netty.buffer.PoolArena$DirectArena.newChunk(PoolArena.java:693)
	at io.netty.buffer.PoolArena.allocateNormal(PoolArena.java:213)
	at io.netty.buffer.PoolArena.tcacheAllocateSmall(PoolArena.java:178)
	at io.netty.buffer.PoolArena.allocate(PoolArena.java:135)
	at io.netty.buffer.PoolArena.allocate(PoolArena.java:127)
	at io.netty.buffer.PooledByteBufAllocator.newDirectBuffer(PooledByteBufAllocator.java:403)
	at io.netty.buffer.AbstractByteBufAllocator.directBuffer(AbstractByteBufAllocator.java:188)
	at io.netty.buffer.AbstractByteBufAllocator.directBuffer(AbstractByteBufAllocator.java:179)
	at io.quarkus.vertx.utils.AppendBuffer.append(AppendBuffer.java:136)
	at io.quarkus.vertx.utils.VertxOutputStream.write(VertxOutputStream.java:178)
	at java.base/java.io.InputStream.transferTo(InputStream.java:797)
	at java.base/sun.nio.ch.ChannelInputStream.transferTo(ChannelInputStream.java:268)
	at org.jboss.resteasy.reactive.common.providers.serialisers.InputStreamMessageBodyHandler.writeTo(InputStreamMessageBodyHandler.java:38)
	at org.jboss.resteasy.reactive.server.providers.serialisers.ServerInputStreamMessageBodyHandler.writeResponse(ServerInputStreamMessageBodyHandler.java:46)
	at org.jboss.resteasy.reactive.server.providers.serialisers.ServerInputStreamMessageBodyHandler.writeResponse(ServerInputStreamMessageBodyHandler.java:17)
	at org.jboss.resteasy.reactive.server.core.ServerSerialisers.invokeWriter(ServerSerialisers.java:218)
	... 12 more
```

in both cases, I see `VertxOutputStream` + `AppendBuffer` involved: should I rename the ticket then?

--- a/independent-projects/vertx-utils/src/main/java/io/quarkus/vertx/utils/VertxOutputStream.java
+++ b/independent-projects/vertx-utils/src/main/java/io/quarkus/vertx/utils/VertxOutputStream.java
@@ -1,11 +1,11 @@
 package io.quarkus.vertx.utils;
 
-import java.io.ByteArrayOutputStream;
 import java.io.IOException;
 import java.io.InterruptedIOException;
 import java.io.OutputStream;
 import java.util.Optional;
 
+import io.vertx.core.Context;
 import org.jboss.logging.Logger;
 
 import io.netty.buffer.ByteBuf;
@@ -13,11 +13,9 @@ import io.netty.handler.codec.http.HttpHeaderNames;
 import io.netty.handler.codec.http.HttpResponse;
 import io.vertx.core.AsyncResult;
 import io.vertx.core.Handler;
-import io.vertx.core.Vertx;
 import io.vertx.core.buffer.Buffer;
 import io.vertx.core.http.HttpServerRequest;
 import io.vertx.core.http.HttpServerResponse;
-import io.vertx.core.http.impl.HttpServerRequestInternal;
 
 /**
  * An {@link OutputStream} forwarding the bytes to Vert.x Web {@link HttpResponse}.
@@ -35,9 +33,7 @@ public class VertxOutputStream extends OutputStream {
     private boolean committed;
     private boolean closed;
     private boolean waitingForDrain;
-    private boolean first = true;
     private Throwable throwable;
-    private ByteArrayOutputStream overflow;
 
     public VertxOutputStream(VertxJavaIoContext context) {
         this.context = context;
@@ -87,29 +83,13 @@ public class VertxOutputStream extends OutputStream {
         //do all this in the same lock
         synchronized (request.connection()) {
             try {
-                boolean bufferRequired = awaitWriteable() || (overflow != null && overflow.size() > 0);
-                if (bufferRequired) {
-                    //just buffer everything
-                    if (overflow == null) {
-                        overflow = new ByteArrayOutputStream();
+                awaitWriteable();
+                if (last) {
+                    if (!response.ended()) { // can happen when an exception occurs during JSON serialization with Jackson
+                        response.end(createBuffer(data), null);
                     }
-                    if (data.hasArray()) {
-                        overflow.write(data.array(), data.arrayOffset() + data.readerIndex(), data.readableBytes());
-                    } else {
-                        data.getBytes(data.readerIndex(), overflow, data.readableBytes());
-                    }
-                    if (last) {
-                        closed = true;
-                    }
-                    data.release();
                 } else {
-                    if (last) {
-                        if (!response.ended()) { // can happen when an exception occurs during JSON serialization with Jackson
-                            response.end(createBuffer(data), null);
-                        }
-                    } else {
-                        response.write(createBuffer(data), null);
-                    }
+                    response.write(createBuffer(data), null);
                 }
             } catch (Exception e) {
                 if (data != null && data.refCnt() > 0) {
@@ -120,13 +100,11 @@ public class VertxOutputStream extends OutputStream {
         }
     }
 
-    private boolean awaitWriteable() throws IOException {
-        if (Vertx.currentContext() == ((HttpServerRequestInternal) request).context()) {
-            return false; // we are on the (right) event loop, so we can write - Netty will do the right thing.
-        }
-        if (first) {
-            first = false;
-            return false;
+    private void awaitWriteable() throws IOException {
+        // is it running in an event loop?
+        if (Context.isOnEventLoopThread()) {
+            // NEVER block the event loop!
+            return;
         }
         assert Thread.holdsLock(request.connection());
         while (response.writeQueueFull()) {
@@ -136,7 +114,6 @@ public class VertxOutputStream extends OutputStream {
             if (response.closed()) {
                 throw new IOException("Connection has been closed");
             }
-            //            registerDrainHandler();
             try {
                 waitingForDrain = true;
                 request.connection().wait();
@@ -146,7 +123,6 @@ public class VertxOutputStream extends OutputStream {
                 waitingForDrain = false;
             }
         }
-        return false;
     }
 
     /**
@@ -257,16 +233,6 @@ public class VertxOutputStream extends OutputStream {
                 synchronized (out.request.connection()) {
                     if (out.waitingForDrain) {
                         out.request.connection().notifyAll();
-                    }
-                    if (out.overflow != null) {
-                        if (out.overflow.size() > 0) {
-                            if (out.closed) {
-                                out.response.end(Buffer.buffer(out.overflow.toByteArray()), null);
-                            } else {
-                                out.response.write(Buffer.buffer(out.overflow.toByteArray()), null);
-                            }
-                            out.overflow.reset();
-                        }
                     }
                 }
             }

